crates/ty_python_semantic/src/types/call/bind.rs

```diff
@@ -1085,23 +1085,152 @@ impl<'db> CallableBinding<'db> {
         // before checking.
         let arguments = arguments.with_self(self.bound_type);
 
-        // TODO: This checks every overload. In the proposed more detailed call checking spec [1],
-        // arguments are checked for arity first, and are only checked for type assignability against
-        // the matching overloads. Make sure to implement that as part of separating call binding into
-        // two phases.
-        //
-        // [1] https://typing.python.org/en/latest/spec/overload.html#overload-call-evaluation
         for overload in &mut self.overloads {
             overload.match_parameters(arguments.as_ref(), argument_forms, conflicting_forms);
         }
     }
 
     fn check_types(&mut self, db: &'db dyn Db, argument_types: &CallArgumentTypes<'_, 'db>) {
+        /// Represents the snapshot of the overload bindings.
+        struct OverloadSnapshots<'db>(SmallVec<[BindingSnapshot<'db>; 1]>);
```

Comment:
It doesn't look like to have to do any fancy merging of these snapshots, they just exist to mark a state to return back to. Could you achieve the same thing by just cloning `self`? There are some extra fields in `CallableBinding` and `Binding` that you'd end up copying, which your snapshot types don't include. But I don't think those fields are substantially larger than the parts that you're already copying in the snapshots, and it might simplify things quite a bit.

Reply:
We _might_ have to do some kind of "merge" logic to determine the final state of the bindings when all the argument lists after expanding one or more types evaluates successfully. Refer to https://github.com/astral-sh/ruff/pull/18382#discussion_r2120026485 comment.

Reply:
> We _might_ have to do some kind of "merge" logic to determine the final state of the bindings when all the argument lists after expanding one or more types evaluates successfully

Make sense! We're not merging yet, but we might need to, so set the types up now to allow that :+1: